File: examples/space-airlock/csharp/src/SpaceAirlock.Steps/SpaceAirlockSteps.cs
using System;
using Xunit;
using SpaceAirlock;

namespace SpaceAirlock.Steps
{
public sealed class SpaceAirlockSteps
{
    private readonly SpaceAirlock _airlock = new();

    public void CrewMemberInside(string name) => _airlock.CrewMemberInside(name);
    public void VisitorInside() => _airlock.VisitorInside();
    public void WearingSuit(string name) => _airlock.WearingSuit(name);
    public void NotWearingSuit(string name) => _airlock.NotWearingSuit(name);
    public void ValidBadge(string name) => _airlock.ValidBadge(name);
    public void VisitorInvalidBadge() => _airlock.VisitorInvalidBadge();
    public void EmergencyOverrideEngaged() => _airlock.EngageEmergencyOverride();
    public void RequestExit(string name) => _airlock.RequestExit(name);
    public void DepressurizationCommanded() => _airlock.Depressurize();
    public void RepressurizationCommanded() => _airlock.Repressurize();

    public void DoorState(string door, string state)
        => _airlock.SetDoorState(ParseDoor(door), ParseDoorStateValue(state));

    public void ChamberPressure(string pressure)
        => _airlock.SetPressure(ParsePressure(pressure));

    public void OpenDoor(string door)
        => _airlock.CommandOpenDoor(ParseDoor(door));

    public void ChamberShouldDepressurize()
        => Assert.Equal(Pressure.Depressurized, _airlock.Pressure);

    public void ChamberShouldRemain(string pressure)
        => Assert.Equal(ParsePressure(pressure), _airlock.Pressure);

    public void DoorShouldUnlock(string door)
        => Assert.Equal(DoorLock.Unlocked, _airlock.DoorLockFor(ParseDoor(door)));

    public void DoorShouldRemainLocked(string door)
        => Assert.Equal(DoorLock.Locked, _airlock.DoorLockFor(ParseDoor(door)));

    public void RequestShouldBeDenied()
        => Assert.True(_airlock.RequestDenied);

    public void SystemShouldDisplay(string message)
        => Assert.Equal(message, _airlock.Message);

    public void AirlockStatusShouldBe(string status)
        => Assert.Equal(status, _airlock.Status);

    private static Door ParseDoor(string door) => door switch
    {
        "inner" => Door.Inner,
        "outer" => Door.Outer,
        _ => throw new ArgumentOutOfRangeException(nameof(door), door, "Unknown door."),
    };

    private static Pressure ParsePressure(string pressure) => pressure switch
    {
        "pressurized" => Pressure.Pressurized,
        "depressurized" => Pressure.Depressurized,
        _ => throw new ArgumentOutOfRangeException(nameof(pressure), pressure, "Unknown pressure."),
    };

    private static DoorStateValue ParseDoorStateValue(string state) => state switch
    {
        "open" => new DoorStateValue.Position(DoorPosition.Open),
        "closed" => new DoorStateValue.Position(DoorPosition.Closed),
        "locked" => new DoorStateValue.Lock(DoorLock.Locked),
        "unlocked" => new DoorStateValue.Lock(DoorLock.Unlocked),
        _ => throw new ArgumentOutOfRangeException(nameof(state), state, "Unknown door state."),
    };
}
}

File: examples/space-airlock/ruby/lib/space_airlock_steps.rb
# frozen_string_literal: true
#
# SpaceAirlockSteps — a thin RSpec helper module that exposes each
# SpaceAirlock instance method as a top-level method delegating to
# `subject`. Including this in a `describe` block lets generated specs
# call `crew_member_inside('alice')` instead of
# `subject.crew_member_inside('alice')`, keeping the `it` blocks dirt
# simple.

require_relative 'space_airlock'

module SpaceAirlockSteps
  # Given helpers
  def crew_member_inside(name);         subject.crew_member_inside(name);         end
  def visitor_inside;                   subject.visitor_inside;                   end
  def wearing_suit(name);               subject.wearing_suit(name);               end
  def not_wearing_suit(name);           subject.not_wearing_suit(name);           end
  def valid_badge(name);                subject.valid_badge(name);                end
  def visitor_invalid_badge;            subject.visitor_invalid_badge;            end
  def door_state(door, state);          subject.door_state(door, state);          end
  def chamber_pressure(pressure);       subject.chamber_pressure(pressure);       end
  def emergency_override_engaged;       subject.emergency_override_engaged;       end

  # When helpers
  def request_exit(name);               subject.request_exit(name);               end
  def open_door(door);                  subject.open_door(door);                  end
  def depressurization_commanded;       subject.depressurization_commanded;       end
  def repressurization_commanded;       subject.repressurization_commanded;       end

  # Then helpers
  def chamber_should_depressurize;      subject.chamber_should_depressurize;      end
  def chamber_should_remain(pressure);  subject.chamber_should_remain(pressure);  end
  def door_should_unlock(door);         subject.door_should_unlock(door);         end
  def door_should_remain_locked(door);  subject.door_should_remain_locked(door);  end
  def request_should_be_denied;         subject.request_should_be_denied;         end
  def system_should_display(message);   subject.system_should_display(message);   end
  def airlock_status_should_be(status); subject.airlock_status_should_be(status); end
end
